src/test/librbd/test_librbd.cc

```diff
@@ -2574,6 +2574,519 @@ TEST_F(TestLibRBD, TestDataPoolIO)
   rados_ioctx_destroy(ioctx);
 }
 
+TEST_F(TestLibRBD, TestCompareAndWriteMismatch)
+{
+  rados_ioctx_t ioctx;
+  rados_ioctx_create(_cluster, m_pool_name.c_str(), &ioctx);
+
+  rbd_image_t image;
+  int order = 0;
+  std::string name = get_temp_image_name();
+  uint64_t size = 20 << 20; /* 20MiB */
+  off_t off = 512;
+
+  ASSERT_EQ(0, create_image(ioctx, name.c_str(), size, &order));
+  ASSERT_EQ(0, rbd_open(ioctx, name.c_str(), &image, NULL));
+
+  // We only support to compare and write the same amount of (len) bytes
+  std::string cmp_buffer("This is a test");
+  std::string write_buffer("Write this !!!");
+  std::string mismatch_buffer("This will fail");
+  std::string read_buffer(cmp_buffer.length(), '1');
+
+  ssize_t written = rbd_write(image, off, cmp_buffer.length(),
+                              cmp_buffer.c_str());
+  ASSERT_EQ(cmp_buffer.length(), written);
+
+  // Compare should fail because of mismatch
+  uint64_t mismatch_off;
+  written = rbd_compare_and_write(image, off, write_buffer.length(),
+                                  mismatch_buffer.c_str(), write_buffer.c_str(),
+                                  &mismatch_off, 0);
+  ASSERT_EQ(-EILSEQ, written);
+  ASSERT_EQ(5U, mismatch_off);
+
+  // check nothing was written
+  ssize_t read = rbd_read(image, off, read_buffer.length(), &read_buffer[0]);
```

Comment:
I changed it for the cases where functions where called like this but I would keep &..[X] for these (if it makes sense to you) because I think it is easier to read:
`std::string cmp_buffer("This is a test");
    ceph::bufferlist cmp_bl;
    cmp_bl.append(&cmp_buffer[0], 5);
    cmp_bl.append(&cmp_buffer[5], 3);
    cmp_bl.append(&cmp_buffer[8], 2);
    cmp_bl.append(&cmp_buffer[10], 4);`

Reply:
Sure, makes sense.